### ISSUE TITLE
DynamoDB Source

### ISSUE THREAD

**Issue by psFried** [estuary/connectors#2]
We'd like to have a source connector for dynamodb. We've done a bit of research and thinking on it already, which I'll write up here.

DynamoDB has a feature called [streams](https://docs.amazonaws.cn/en_us/amazondynamodb/latest/developerguide/Streams.html) that allows reading CDC events. There is a separate feature that can automatically forward these CDC events to a Kinesis stream. Our connector should use dynamodb streams, _not_ the kinesis feature. The kinesis feature has a seemingly higher cost, and CDC events may be duplicated in the kinesis stream. Additionally, the kinesis feature does not scale automatically to adjust for the throughput on the dynamodb table. Dynamodb streams has a lower cost and scales automatically as the dynamodb table itself scales. The dynamodb streams API is very similar to the Kinesis API, though, so the two implementations can likely share a lot of code.

When processing CDC data, it's important to process events in the correct order. If you have an insert followed by a delete on the same key, you will end up with incorrect results if you processed the delete before the insert. This is complicated by design of Shards in dynamodb streams. Shards each have a parent shard, and each shard may have 0 or more child shards. You must read a parent shard entirely before you start reading a child shard if you want to process documents in the correct order. This is problematic because of how we distribute reads of the dynamodb (or kinesis) shards among the set of Flow consumer shards. Consumer shard A may decide to process dynamo shard X before Consumer shard B has finished reading the parent of X. Solving this seems like it would require some shared state.

Another possible approach for dealing with ordering is to use the `ApproximateCreationDateTime` of each dynamodb record with a `maximize` reduction strategy on the root document so that the most recent CDC document for each key is the one that will take precedence.

**Comment by williamhbaker** [comment]
I've spent some time working on this and getting a working implementation.

For a single Flow consumer shard it's reasonably straightforward, but becomes more complicated when considering multiple Flow consumer shards for reasons alluded to in the original issue description, in addition to considerations for backfilling existing data from a table.

I'll describe the general mode of operation first, followed by a description of the challenges and a _possible_ way to work with scaled-out Flow consumer shards.

## General Mode of Operation (Single Flow Shard)

DynamoDB allows for a full [scan](https://docs.aws.amazon.com/amazondynamodb/latest/APIReference/API_Scan.html) of a table to backfill pre-existing records, and for reading change event records from a [stream](https://docs.aws.amazon.com/amazondynamodb/latest/developerguide/Streams.html) for that table. Results from a scan are not able to be sorted in the general case - they are _resumable_ from the last scanned key, but the ordering is (apparently) based on how the records are arranged in table partitions according to the hash of the key used internally by Dynamo so they are essentially returned in a random order.

So to start capturing a table, you can do a scan query to backfill the existing records, and start reading the change stream from a time just prior to starting the backfill to pick up changes. It isn't possible to do an exactly-once kind of audit-log capture, but it will be "eventually" correct when materialized as last-write-wins.

The typical strategy for other existing systems is to first scan the entire table, and then start reading stream records from just before the backfill started. DynamoDB streams have a retention limit of 24 hours, so this means that the entire backfill must complete in <24 hours. I've been endeavouring to not have this requirement for our capture (it seems like it could be a major bummer for huge tables), and I believe it can be avoided by reading the stream records simultaneously with doing the backfill scan, so that we stay somewhat "caught up" with the stream and don't lose any stream records due to them aging out.

A few scenarios to consider for reading streaming records concurrently with a backfill scan:

- A change happens to a record we have already scanned, and this change event is picked up from the stream: This is totally fine, and will be reflected in the captured collection as the old record (from the table scan) later followed by the new record (from the stream change event). In fact this is the preferred way for things to happen, and would be used exclusively if it were possible to do so.
- A change happens to a record we haven't already scanned: The record will be scanned from the table at some point in the future, and so will the change event from the stream. There will be a duplicate of the record in the captured collection, which will be resolved with a last-write-wins materialization.
- Multiple changes happen to a record we haven't already scanned: It's possible that the current record would be scanned from the table first, followed by an "older" change event from the change stream and _then_ the change representing the current record from the change stream. Reading the collection sequentially it would look like the record reverted back to an older version before later changing again to the current version. As above, this will eventually be correct with a last-write-wins materialization.

In case it wasn't obvious, this general "weird ordering of records during the backfill stage that eventually corrects with a last-write-wins materialization" would _also_ happen if the entire table were scanned first before reading any of the change events.

There is one unique challenge with the simultaneous stream-reading and backfill-scanning strategy, and it is this: We could read a key from the table during a scan at nearly the same instant as it is changed, resulting in a race between the "backfiller" writing the record to the collection vs. the "change streamer" writing the record to the collection. It is possible that the "old" version of the record could be read by a scan and before it is written to the collection, the change event from the stream with the "new" version of the record is read from the stream & written to the collection, which is then clobbered by the "old" record from the table scan finally being written to the collection. To mitigate this I have mentioned only reading from "closed" stream shards, since we know that no new records will be written to them. This might not even work 100% of the time though since it may be possible for a stream shard to receive an update and then immediately be closed nearly simultaneously with the table scan reading the value for the updated key. A more robust a simpler way to handle is to hold back on reading stream updates that are less than some time-horizon value old (maybe ~15 minutes or mote) using the [`ApproximateCreationDateTime`](https://docs.aws.amazon.com/amazondynamodb/latest/APIReference/API_streams_StreamRecord.html) value that is present in stream records. Per the field's name this is an approximate time but we can be quite conservative in our time horizon while backfilling since we only really need to stay caught up within 24 hours on the stream.

Once the backfill is done and there are no more table scans, the capture can switch to a "real time" streaming mode and not hold back any change events for the rest of its tenure.

## Scale Out

For capturing very large tables we'd like to be able to split Flow task shards to divide up the work of capturing the table, ideally both for the backfill and the on-going streaming during & after that.

The mechanism for splitting up a DynamoDB table scan for multiple threads/processes is via a [parallel scan](https://docs.aws.amazon.com/amazondynamodb/latest/developerguide/Scan.html#Scan.ParallelScan). A number of `TotalSegments` for the query is set, and a separate worker can handle each segment. This divides the table's hash space in into multiple parts, allowing each one to be handled separately in parallel. Conceivably each Flow task shard could handle a portion of the segments for the scan.

DynamoDB streams work similarly to Kinesis, with some important differences. Most notably, there is no way to tell which keys a stream shard contains - there is no [HashKeyRange](https://docs.aws.amazon.com/kinesis/latest/APIReference/API_HashKeyRange.html) like Kinesis shards have. Stream shards have IDs like `shardId-00000001688558464086-cdfd95fe`. [Indications are](https://stackoverflow.com/questions/44266633/how-do-dynamodb-streams-distribute-records-to-shards) that shards correspond to partitions of a table, and that a given key will exist in only one active shard. So DynamoDB stream shard processing work could be scaled out to multiple Flow task shards based on a hash of the stream shard ID so that a stream shard (well, shard tree really, but I'm omitting some of the nitty-gritty details, see mentions of reading parent shards first in the original issue description) is processed by a single Flow task shard.

To recap:

- Backfill scans can be divided into segments, and different segments can be processed by different workers.
- Individual stream shards (shard trees actually, see above) can be processed by different workers.
- There is no way to tell what key range a backfill segment will contain, but a given key will only occur in one segment.
- There is no way to tell what key range a stream shard will contain, but a given key will only occur in one active stream shard.
- There is also no way to scan a table partition directly, although this is less consequential since there is _also_ no way to tell which partition that a stream shard corresponds to that I am aware of.
- The streaming mode of operation needs to work differently while backfilling is still in progress: It must hold back emitting change events based on a time-horizon to avoid a race condition of updates with backfill scans, but can start operating in real-time when backfill scans are complete.

For illustrative purposes, here a couple of ways this could function really nicely if DynamoDB worked a little differently, but are currently impossible:

- It would be nice to map DynamoDB table partitions to Flow task shards, where each Flow task shard were responsible for backfilling certain DynamoDB table partitions and their corresponding DynamoDB stream shards. We can't do this since we can't specificy which table partitions to scan for a backfill query, and also we can't tell which shards correspond to which partitions.
- It would be nice to map key ranges to Flow task shards. But we can't tell what the key range of any particular DynamoDB stream shard is, among other things.

What could be possible is for each Flow task shard to process some of the backfill scan segements, and also some of the DynamoDB stream shards, with the key ranges seen in the scan segements being different than the key ranges from the stream shards. With this strategy, a Flow task shard can only switch to real time streaming once _all_ backfill scan segments are complete - not only the ones it is responsible for - since its streaming of a key range may race the on-going backfill scan from another shard that is handling the segments for that key range. This implies some mechanism of coordination between Flow task shards, allowing them to know collectively that they are all done with the backfill. There is some precedent for creating and using a DynamoDB table for tracking consumer state (see KCL [leases table](https://aws.amazon.com/blogs/big-data/processing-amazon-dynamodb-streams-using-the-amazon-kinesis-client-library/)), so this might be an option for coordination between Flow task shards.

## Other Considerations

- Backfilling from a table using a scaled-out, as fast as possible strategy might not always be desirable, since it could consume all of the RCU's of a table and impact other queries in an undesirable way. Using the [ConsumedCapacity](https://docs.aws.amazon.com/amazondynamodb/latest/APIReference/API_Scan.html#API_Scan_ResponseSyntax) information returned from a scan we could allow for configuring how much RCU a capture is allowed to use, but only for an individual Flow task shard. I'm on the fence about implementing this immediately but I'll speculate and say that it is probably something that will be requested before too long.
- There is a way to take a snapshot of an entire DynamoDB table and save it to an S3 bucket in a JSON-approximate form. The resulting JSON files could be parsed and ingested without using RCU on the main table, and probably a lot faster than a scan query. We wouldn't be able to backfill and stream simultaneously, so there would be a requirement that the entire snapshot can be read in <24 hours. See [an example](https://rockset.com/blog/20x-faster-ingestion-with-rocksets-new-dynamodb-connector/) describing this. I see this as a possible future alternate option for backfilling to avoid RCU usage, but not as the only possible option since it brings more complexity to the setup that probably wouldn't be needed for most smaller tables.
